### PR TITLE
Initial optimizations on PrepareProposal/ProcessProposal for EL block production with Engine API 

### PR DESCRIPTION
**Pull request by Raneet10**
# Description

This PR is a first attempt to pipeline the EL payload production via the ABCI stages (Prepare/ProcessProposal mostly). The results saw a good improvement in `PrepareProposal` stage but needs to be tested for various other corner cases. 

Before
```
12:20PM INF 🕒 End PrepareProposal: duration=88.973923ms height=30 module=server momentTime=20:34.164467 payloadSize=296102
```

After
```
12:08PM INF 🕒 End PrepareProposal: duration=1.594101ms height=53 module=server momentTime=08:50.719651 payloadSize=292562
```

# Changes

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] Changes only for a subset of nodes

# Checklist

- [x] I have added at least 2 reviewer or the whole pos-v1 team
- [x] I have added sufficient documentation in code
- [x] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply

## Testing

- [ ] I have added unit tests
- [ ] I have added tests to CI
- [x] I have tested this code manually on local environment
- [ ] I have tested this code manually on remote devnet using express-cli
- [ ] I have tested this code manually on amoy/mumbai
- [ ] I have created new e2e tests into express-cli

### CODE REVIEW

File: app/abci.go [hunk 1]
```diff
@@ -236,23 +251,57 @@ func (app *HeimdallApp) NewProcessProposalHandler() sdk.ProcessProposalHandler {
...
 			return &abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_REJECT}, nil
 		}
 
+		// app.execPayloadCache.Add(payload.LatestValidHash, payload)
 		duration := time.Since(start)
 		formatted := fmt.Sprintf("%.6fms", float64(duration)/float64(time.Millisecond))
-		logger.Info("🕒 End ProcessProposal:", "duration", formatted, "height", req.Height, "payloadSize", len(req.Txs[0]), "momentTime", time.Now().Format("04:05.000000"))
+		logger.Info("🕒 ProcessProposal:" + formatted)
+
+		app.nextBlockChan <- nextELBlockCtx{height: req.Height + 1,
```
Comment (avalkov) [inline app/abci.go:297]: Where is `nextExecPayload` read from?

File: app/abci.go [hunk 2]
```diff
@@ -236,23 +251,57 @@ func (app *HeimdallApp) NewProcessProposalHandler() sdk.ProcessProposalHandler {
 		var executionPayload engine.ExecutionPayload
 		err = json.Unmarshal(metadata.MarshaledExecutionPayload, &executionPayload)
 		if err != nil {
+			app.currBlockChan <- nextELBlockCtx{height: req.Height,
```
Comment (avalkov) [inline app/abci.go:254]: Why we don't simply return `app.currBlockChan <- nextELBlockCtx{height: req.Height, context: ctx}` in these cases ?

File: app/abci.go [hunk 3]
```diff
@@ -58,51 +58,56 @@ func (app *HeimdallApp) NewPrepareProposalHandler() sdk.PrepareProposalHandler {
 		}
 
 		// Engine API
-		latestBlock, err := app.caller.BorChainClient.BlockByNumber(ctx, big.NewInt(req.Height-1)) // change this to a keeper
-		if err != nil {
-			return nil, err
-		}
+		var payload *engine.Payload
+		if app.latestExecPayload != nil {
+			payload = app.latestExecPayload
```
[the remark this answers is not in the document]
Comment (Raneet10) [reply app/abci.go:63]: Yes good point. Another option is to set `app.latestExecPayload` to nil once we use it but the height check approach seems better.

File: app/abci.go (hunk 2 above)
Comment (Raneet10) [reply app/abci.go:254]: Saves a bor RPC call.

Comment (avalkov) [reply app/abci.go:260]: What will be consequences if the ParentHash is invalid? Proposed by malicious or corrupted proposer.

Comment (Raneet10) [reply app/abci.go:260]: Actually we can use the last block value stored in the keeper.

Comment (avalkov) [reply app/abci.go:260]: Ideally thats what we should use, instead of fetching latest block number from erigon. Fetching of the latest block number from erigon was just a temporary workaround.
